# BÀI 1: LÀM QUEN VỚI PYTORCH

## 1. Mục tiêu

- Sau khi hoàn thành bài thực hành, sinh viên có thể:
- Tạo, kiểm tra và thao tác với Tensor trong PyTorch.
- Thực hiện indexing, slicing, reshape và các phép toán ma trận trên Tensor.
- Sử dụng các hàm thống kê cơ bản trên Tensor.
- Hiểu cơ chế automatic differentiation (autograd) và vai trò của gradient.
- Sử dụng optimizer để giải bài toán tối ưu đơn giản bằng Gradient Descent.
- PyTorch cung cấp hai khả năng nền tảng: tính toán Tensor (tương tự NumPy, có thể tăng tốc bằng GPU) và automatic differentiation để huấn luyện mô hình.

## 2. Bài tập thực hành

### Bài 1. Tạo Tensor cơ bản

Tạo một Tensor số nguyên kích thước 2 × 3 và hiển thị kết quả.

**Yêu cầu:** Cho biết kích thước, số chiều và kiểu dữ liệu của Tensor `a`.

In [1]:
import torch

a = torch.tensor(
    [[1, 2, 3],
     [4, 5, 6]],
    dtype=torch.int32
)

print(a)
print("Kích thước:", a.shape)
print("Số chiều:", a.ndim)
print("Kiểu dữ liệu:", a.dtype)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Kích thước: torch.Size([2, 3])
Số chiều: 2
Kiểu dữ liệu: torch.int32


#### Nhận xét

Tensor `a` có kích thước **2 × 3**, gồm **2 chiều** và có kiểu dữ liệu **torch.int32**.

- `shape` cho biết kích thước của Tensor.
- `ndim` cho biết số chiều của Tensor.
- `dtype` cho biết kiểu dữ liệu của các phần tử.

Như vậy, Tensor `a` đáp ứng đúng yêu cầu là một Tensor số nguyên có 2 hàng và 3 cột.

### Bài 2. Tạo các Tensor thường dùng

Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị.

**Yêu cầu:**
- Nhận xét sự khác nhau giữa `torch.rand()`và `torch.randn()`.
- Tạo thêm một Tensor 5 × 5 chứa toàn giá trị 10.

In [4]:
import torch

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
f = torch.full((5, 5), 10)

print("Tensor a:")
print(a)

print("\nTensor b:")
print(b)

print("\nTensor c:")
print(c)

print("\nTensor d:")
print(d)

print("\nTensor e:")
print(e)

print("\nTensor f:")
print(f)

Tensor a:
tensor([[0.8214, 0.2383, 0.5161, 0.7890],
        [0.8385, 0.3500, 0.1080, 0.8223],
        [0.7860, 0.9607, 0.8596, 0.7621]])

Tensor b:
tensor([[-0.0458,  0.6343, -0.3945, -0.7891],
        [ 1.3546, -2.0543,  0.8749, -2.9584],
        [-0.0516,  0.8914,  0.6150,  1.3856]])

Tensor c:
tensor([[0., 0., 0.],
        [0., 0., 0.]])

Tensor d:
tensor([[1., 1., 1.],
        [1., 1., 1.]])

Tensor e:
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

Tensor f:
tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


#### Nhận xét

- `torch.rand()` tạo Tensor chứa các giá trị ngẫu nhiên trong khoảng **[0, 1)**.
- `torch.randn()` tạo Tensor chứa các giá trị ngẫu nhiên theo **phân phối chuẩn**, vì vậy có thể xuất hiện cả giá trị âm và giá trị dương.
- `torch.zeros()` tạo Tensor chứa toàn giá trị 0.
- `torch.ones()` tạo Tensor chứa toàn giá trị 1.
- `torch.eye()` tạo ma trận đơn vị.
- `torch.full((5, 5), 10)` tạo Tensor kích thước **5 × 5** chứa toàn giá trị 10.

Do `torch.rand()` và `torch.randn()` sinh dữ liệu ngẫu nhiên nên kết quả có thể thay đổi sau mỗi lần chạy chương trình.

### Bài 3. Kiểm tra thông tin Tensor

Sử dụng `shape`, `size()`, `ndim` và `dtype` để kiểm tra các thuộc tính của Tensor.

**Yêu cầu:** Giải thích ý nghĩa của `torch.Size([2, 3, 4])`.

In [10]:
import torch

a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


#### Nhận xét

Tensor `a` có kích thước `torch.Size([2, 3, 4])`, gồm **3 chiều**.

- Chiều thứ nhất có kích thước 2.
- Chiều thứ hai có kích thước 3.
- Chiều thứ ba có kích thước 4.
- Tensor có tổng cộng `2 × 3 × 4 = 24` phần tử.
- `shape` và `size()` đều cho biết kích thước của Tensor.
- `ndim` cho biết số chiều của Tensor.
- `dtype` cho biết kiểu dữ liệu của các phần tử; trong trường hợp này là `torch.float32`.

Vì vậy, `torch.Size([2, 3, 4])` có thể hiểu là Tensor gồm **2 khối**, mỗi khối có **3 hàng và 4 cột**.

### Bài 4. Slicing và Indexing

Truy xuất một phần dữ liệu từ Tensor nhiều chiều.

**Yêu cầu:**
- Cho biết shape của `a[1]` và `a[1:, 2:4]`.
- Thử lấy hai phần tử cuối cùng trên chiều thứ ba.

In [8]:
import torch

a = torch.randn(3, 4, 5)

print("Original tensor:")
print(a)

print("\nTensor a[1]:")
print(a[1])

print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])

print("\nShape of a[1]:", a[1].shape)
print("Shape of a[1:, 2:4]:", a[1:, 2:4].shape)

print("\nHai phần tử cuối trên chiều thứ ba:")
print(a[:, :, -2:])

Original tensor:
tensor([[[-1.2256, -1.8925, -0.5691, -0.1997, -0.4566],
         [-0.5549,  1.9534, -1.7504, -2.2180,  0.1963],
         [ 0.7355,  0.4445,  2.4406, -0.5319,  1.4445],
         [ 0.0169,  1.3417, -0.3366,  2.1456, -0.3936]],

        [[-0.9564, -1.1832,  0.8414,  1.2315, -0.1398],
         [ 0.1743,  0.4252, -0.3746, -0.3717,  1.1844],
         [ 1.6670, -0.5791,  0.3283, -1.5691,  1.8369],
         [ 0.3419,  0.4785, -0.2754, -0.4597,  1.3616]],

        [[-1.1714,  0.6019, -0.2449,  0.6523, -0.5038],
         [ 0.1796,  0.0191, -0.7046, -0.2266,  1.7307],
         [-1.3221, -0.7693, -0.4230,  0.8929,  0.0484],
         [-0.7146,  0.0891,  1.6569, -0.9753, -0.3526]]])

Tensor a[1]:
tensor([[-0.9564, -1.1832,  0.8414,  1.2315, -0.1398],
        [ 0.1743,  0.4252, -0.3746, -0.3717,  1.1844],
        [ 1.6670, -0.5791,  0.3283, -1.5691,  1.8369],
        [ 0.3419,  0.4785, -0.2754, -0.4597,  1.3616]])

Tensor a[1:, 2:4]:
tensor([[[ 1.6670, -0.5791,  0.3283, -1.5691,  1.8

#### Nhận xét

Tensor ban đầu `a` có kích thước `3 × 4 × 5`.

- `a[1]` lấy phần tử có chỉ số 1 trên chiều thứ nhất, tức khối thứ hai của Tensor. Kết quả có shape `torch.Size([4, 5])`.
- `a[1:, 2:4]` lấy các phần tử từ chỉ số 1 trở đi trên chiều thứ nhất và từ chỉ số 2 đến trước chỉ số 4 trên chiều thứ hai. Kết quả có shape `torch.Size([2, 2, 5])`.
- `a[:, :, -2:]` giữ nguyên hai chiều đầu và lấy hai phần tử cuối cùng trên chiều thứ ba. Kết quả có shape `torch.Size([3, 4, 2])`.

Qua đó, indexing dùng để truy xuất phần tử hoặc một phần cụ thể của Tensor, còn slicing cho phép lấy một phạm vi dữ liệu theo từng chiều.

### Bài 5. Reshape và Flatten

Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử.

**Yêu cầu:**
- Kiểm tra số phần tử trước và sau khi reshape.
- Thử reshape Tensor thành kích thước 2 × 6.

In [9]:
import torch

a = torch.randn(3, 4)

print("Original:")
print(a)

print("\nFlatten:")
print(a.ravel())

print("\nReshape to 3 x 2 x 2:")
b = a.reshape(3, 2, 2)
print(b)

print("\nSố phần tử ban đầu:", a.numel())
print("Số phần tử sau khi reshape:", b.numel())

print("\nReshape to 2 x 6:")
c = a.reshape(2, 6)
print(c)

print("\nShape ban đầu:", a.shape)
print("Shape sau khi reshape 3 x 2 x 2:", b.shape)
print("Shape sau khi reshape 2 x 6:", c.shape)

Original:
tensor([[-2.8591,  0.2357,  0.7638,  0.7378],
        [-0.6713, -0.0088, -2.4365,  0.3609],
        [-1.8911, -0.3104,  2.8299,  1.0460]])

Flatten:
tensor([-2.8591,  0.2357,  0.7638,  0.7378, -0.6713, -0.0088, -2.4365,  0.3609,
        -1.8911, -0.3104,  2.8299,  1.0460])

Reshape to 3 x 2 x 2:
tensor([[[-2.8591,  0.2357],
         [ 0.7638,  0.7378]],

        [[-0.6713, -0.0088],
         [-2.4365,  0.3609]],

        [[-1.8911, -0.3104],
         [ 2.8299,  1.0460]]])

Số phần tử ban đầu: 12
Số phần tử sau khi reshape: 12

Reshape to 2 x 6:
tensor([[-2.8591,  0.2357,  0.7638,  0.7378, -0.6713, -0.0088],
        [-2.4365,  0.3609, -1.8911, -0.3104,  2.8299,  1.0460]])

Shape ban đầu: torch.Size([3, 4])
Shape sau khi reshape 3 x 2 x 2: torch.Size([3, 2, 2])
Shape sau khi reshape 2 x 6: torch.Size([2, 6])


#### Nhận xét

Tensor ban đầu có kích thước `3 × 4`, tương ứng với **12 phần tử**.

- `a.ravel()` chuyển Tensor thành Tensor một chiều có 12 phần tử.
- `a.reshape(3, 2, 2)` thay đổi Tensor thành kích thước `3 × 2 × 2` nhưng vẫn giữ nguyên 12 phần tử.
- `a.reshape(2, 6)` thay đổi Tensor thành kích thước `2 × 6` và cũng giữ nguyên 12 phần tử.
- `numel()` cho biết tổng số phần tử của Tensor.

Như vậy, thao tác reshape chỉ thay đổi hình dạng của Tensor mà không làm thay đổi tổng số phần tử.

### Bài 6. Phép toán và nhân ma trận

Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử `@`.

**Yêu cầu:**
- Phân biệt phép toán `a * b` với `a @ b.T`.
- Cho biết shape của kết quả `a @ b.T`.

In [11]:
import torch

a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:")
print(a)

print("\nb:")
print(b)

print("\na * b:")
print(a * b)

print("\na + b:")
print(a + b)

print("\na / b:")
print(a / b)

print("\na^2:")
print(a ** 2)

print("\nMatrix multiplication:")
print(a @ b.T)

print("\nShape của a * b:", (a * b).shape)
print("Shape của a @ b.T:", (a @ b.T).shape)

a:
tensor([[-0.8296,  0.3671,  0.3164],
        [-0.6907,  1.5039,  0.7566]])

b:
tensor([[ 1.8318,  0.9836, -1.0823],
        [ 1.3740, -1.4151, -1.9513]])

a * b:
tensor([[-1.5196,  0.3611, -0.3425],
        [-0.9490, -2.1282, -1.4763]])

a + b:
tensor([[ 1.0022,  1.3507, -0.7659],
        [ 0.6834,  0.0889, -1.1947]])

a / b:
tensor([[-0.4529,  0.3732, -0.2923],
        [-0.5027, -1.0628, -0.3877]])

a^2:
tensor([[0.6882, 0.1348, 0.1001],
        [0.4770, 2.2619, 0.5724]])

Matrix multiplication:
tensor([[-1.5010, -2.2767],
        [-0.6047, -4.5535]])

Shape của a * b: torch.Size([2, 3])
Shape của a @ b.T: torch.Size([2, 2])


#### Nhận xét

- `a * b` thực hiện phép nhân theo từng phần tử tương ứng của hai Tensor. Vì `a` và `b` đều có kích thước `2 × 3`, kết quả cũng có shape `torch.Size([2, 3])`.
- `a @ b.T` thực hiện phép nhân ma trận. Tensor `b` có kích thước `2 × 3`, sau khi chuyển vị `b.T` có kích thước `3 × 2`.
- Do đó, phép nhân giữa `a` có kích thước `2 × 3` và `b.T` có kích thước `3 × 2` tạo ra Tensor có shape `torch.Size([2, 2])`.
- Như vậy, toán tử `*` dùng để nhân theo từng phần tử, còn toán tử `@` dùng để thực hiện phép nhân ma trận.

### Bài 7. Các hàm thống kê trên Tensor

Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột.

**Yêu cầu:** Thay `dim=0` bằng `dim=1` và nhận xét sự khác nhau.

In [12]:
import torch

a = torch.randn(3, 4)

print("Tensor a:")
print(a)

print("\n--- dim = 0 ---")

print("\nMean:")
print(torch.mean(a, dim=0))

print("\nStandard deviation:")
print(torch.std(a, dim=0))

print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

print("\n--- dim = 1 ---")

print("\nMean:")
print(torch.mean(a, dim=1))

print("\nStandard deviation:")
print(torch.std(a, dim=1))

print("\nCumulative sum:")
print(torch.cumsum(a, dim=1))

Tensor a:
tensor([[ 0.1501, -0.1232,  1.2946,  2.1602],
        [-1.8371, -0.0897, -0.2176, -2.8466],
        [-0.4902,  0.3316,  0.0435,  0.8931]])

--- dim = 0 ---

Mean:
tensor([-0.7257,  0.0396,  0.3735,  0.0689])

Standard deviation:
tensor([1.0143, 0.2535, 0.8083, 2.6032])

Cumulative sum:
tensor([[ 0.1501, -0.1232,  1.2946,  2.1602],
        [-1.6870, -0.2128,  1.0770, -0.6865],
        [-2.1772,  0.1188,  1.1205,  0.2066]])

--- dim = 1 ---

Mean:
tensor([ 0.8704, -1.2477,  0.1945])

Standard deviation:
tensor([1.0566, 1.3299, 0.5769])

Cumulative sum:
tensor([[ 0.1501,  0.0269,  1.3215,  3.4817],
        [-1.8371, -1.9267, -2.1443, -4.9909],
        [-0.4902, -0.1586, -0.1151,  0.7780]])


#### Nhận xét

- Khi `dim=0`, các phép toán được thực hiện theo chiều thứ nhất của Tensor, tức là tính theo từng cột. Với Tensor kích thước `3 × 4`, kết quả của `mean` và `std` có 4 phần tử.
- Khi `dim=1`, các phép toán được thực hiện theo chiều thứ hai, tức là tính theo từng hàng. Kết quả của `mean` và `std` có 3 phần tử.
- `torch.cumsum()` tính tổng tích lũy theo chiều được chỉ định. Với `dim=0`, tổng được tích lũy theo các hàng cho từng cột; với `dim=1`, tổng được tích lũy theo các cột trong từng hàng.

Như vậy, thay đổi giá trị `dim` sẽ làm thay đổi chiều mà phép toán thống kê được thực hiện.

### Bài 8. Autograd cơ bản

Tạo biến có `requires_grad=True` và sử dụng `backward()` để tính đạo hàm tự động.

**Yêu cầu:**
- Tính thủ công đạo hàm của `y = x²` tại `x = 3.6` và so sánh với `x.grad`.
- Đổi hàm thành `y = 3*x**2 + 2*x + 1` và kiểm tra gradient.

In [17]:
import torch

# Trường hợp 1: y = x^2
x = torch.tensor(
    3.6,
    requires_grad=True
)

y = x * x
y.backward()

print("--- Trường hợp 1: y = x^2 ---")
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)


# Trường hợp 2: y = 3*x^2 + 2*x + 1
x = torch.tensor(
    3.6,
    requires_grad=True
)

y = 3 * x**2 + 2 * x + 1
y.backward()

print("\n--- Trường hợp 2: y = 3*x^2 + 2*x + 1 ---")
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

--- Trường hợp 1: y = x^2 ---
x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)

--- Trường hợp 2: y = 3*x^2 + 2*x + 1 ---
x = tensor(3.6000, requires_grad=True)
y = tensor(47.0800, grad_fn=<AddBackward0>)
x.grad = tensor(23.6000)


#### Nhận xét

Với hàm `y = x²`, đạo hàm theo `x` là: `dy/dx = 2x`

Tại `x = 3.6`: `dy/dx = 2 × 3.6 = 7.2`

Kết quả này trùng với giá trị `x.grad` do PyTorch tính bằng `backward()`.

Khi đổi hàm thành `y = 3*x**2 + 2*x + 1`, đạo hàm là: `dy/dx = 6x + 2`

Tại `x = 3.6`: `dy/dx = 6 × 3.6 + 2 = 23.6`

Giá trị này cũng trùng với gradient được PyTorch tính tự động. Qua đó, `autograd` cho phép PyTorch theo dõi các phép toán và tự động tính gradient của Tensor có `requires_grad=True`.

### Bài 9. Gradient Descent tìm hệ số đa thức

Sinh dữ liệu từ đa thức `y = x² + 2x + 3`, sau đó sử dụng NAdam và autograd để tìm lại ba hệ số.

**Yêu cầu:**
- So sánh ba hệ số học được với `[1, 2, 3]`.
- Giải thích vai trò của `optimizer.zero_grad()`, `mse.backward()` và `optimizer.step()`.

In [1]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])

N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)

x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()

    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))

    mse.backward()
    optimizer.step()

print("Learned coefficients:")
print(w)

Initial coefficients:
tensor([[0.6880],
        [0.0690],
        [1.4652]], requires_grad=True)
Learned coefficients:
tensor([[1.0001],
        [2.0003],
        [2.9938]], requires_grad=True)


#### Nhận xét

Ba hệ số ban đầu được khởi tạo ngẫu nhiên là `[0.6880, 0.0690, 1.4652]`. Sau 1000 epoch, mô hình học được các hệ số xấp xỉ `[1.0001, 2.0003, 2.9938]`, rất gần với các hệ số thực `[1, 2, 3]` của đa thức `y = x² + 2x + 3`.

Trong quá trình tối ưu:
- `optimizer.zero_grad()` dùng để xóa gradient của vòng lặp trước, tránh gradient bị cộng dồn.
- `mse.backward()` thực hiện lan truyền ngược để tính gradient của hàm mất mát MSE đối với các hệ số `w`.
- `optimizer.step()` sử dụng các gradient đã tính để cập nhật các hệ số `w` nhằm làm giảm sai số.

Như vậy, thông qua autograd và optimizer NAdam, PyTorch có thể tối ưu các tham số và tìm lại gần đúng các hệ số của đa thức ban đầu.

### Bài 10. Giải hệ phương trình bằng Autograd

Tối thiểu hóa tổng sai số bình phương để tìm `A`, `B`, `C`, `D` thỏa bốn phương trình.

**Yêu cầu:**
- Kiểm tra nghiệm thu được với bốn phương trình: `A+B=9`, `C-D=1`, `A+C=8`, `B-D=2`.
- Nhận xét vì sao bài toán có thể có nhiều nghiệm.

In [2]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000

optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2

    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

print("\nKiểm tra:")
print("A + B =", A + B)
print("C - D =", C - D)
print("A + C =", A + C)
print("B - D =", B - D)

A = tensor(4.2982, requires_grad=True)
B = tensor(4.7017, requires_grad=True)
C = tensor(3.7018, requires_grad=True)
D = tensor(2.7018, requires_grad=True)

Kiểm tra:
A + B = tensor(9.0000, grad_fn=<AddBackward0>)
C - D = tensor(1.0000, grad_fn=<SubBackward0>)
A + C = tensor(8.0000, grad_fn=<AddBackward0>)
B - D = tensor(2.0000, grad_fn=<SubBackward0>)


#### Nhận xét

Sau 2000 epoch, mô hình tìm được nghiệm xấp xỉ `A = 4.2982`, `B = 4.7017`, `C = 3.7018` và `D = 2.7018`.

Kiểm tra lại bốn phương trình:
- `A + B = 9.0000`
- `C - D = 1.0000`
- `A + C = 8.0000`
- `B - D = 2.0000`

Như vậy, các giá trị tìm được thỏa mãn cả bốn phương trình.

Hệ phương trình có thể có nhiều nghiệm vì bốn phương trình không hoàn toàn độc lập. Từ ba phương trình đầu có thể biểu diễn `B = 9 - A`, `C = 8 - A` và `D = 7 - A`. Khi thay vào phương trình `B - D = 2` thì phương trình này luôn được thỏa mãn. Vì vậy, một biến có thể thay đổi tự do và các biến còn lại thay đổi tương ứng mà vẫn thỏa mãn hệ.

## 3. Bài tập áp dụng

### Bài 11. Viết chương trình PyTorch để thực hiện các yêu cầu sau:

**Yêu cầu:**
1. Tạo Tensor `X` có 24 phần tử và reshape lần lượt thành `4 × 6`, `2 × 12` và `2 × 3 × 4`.
2. Tạo hai Tensor `A`, `B` kích thước `3 × 4`; tính `A+B`, `A*B`, `A/B`, trung bình theo cột và độ lệch chuẩn theo cột.
3. Cho `y = 2x² + 5x + 3`. Dùng autograd tính `dy/dx` tại `x = 2` và kiểm tra bằng phép tính thủ công.
4. Thay đa thức trong Bài 9 bằng `y = 2x² - 3x + 5`; huấn luyện và cho biết các hệ số tìm được.

#### Câu 1. Reshape Tensor

Tạo Tensor `X` có 24 phần tử và reshape lần lượt thành `4 × 6`, `2 × 12` và `2 × 3 × 4`.

In [6]:
import torch

X = torch.arange(24)

print("Tensor X:")
print(X)

print("\nX reshape 4 x 6:")
print(X.reshape(4, 6))

print("\nX reshape 2 x 12:")
print(X.reshape(2, 12))

print("\nX reshape 2 x 3 x 4:")
print(X.reshape(2, 3, 4))

Tensor X:
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])

X reshape 4 x 6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

X reshape 2 x 12:
tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

X reshape 2 x 3 x 4:
tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])


#### Câu 2. Phép toán và thống kê

Tạo hai Tensor `A`, `B` kích thước `3 × 4`; tính `A+B`, `A*B`, `A/B`, trung bình theo cột và độ lệch chuẩn theo cột.

In [7]:
A = torch.rand(3, 4)
B = torch.rand(3, 4)

print("Tensor A:")
print(A)

print("\nTensor B:")
print(B)

print("\nA + B:")
print(A + B)

print("\nA * B:")
print(A * B)

print("\nA / B:")
print(A / B)

print("\nTrung bình theo cột của A:")
print(torch.mean(A, dim=0))

print("\nĐộ lệch chuẩn theo cột của A:")
print(torch.std(A, dim=0))

Tensor A:
tensor([[0.2149, 0.0386, 0.7421, 0.7967],
        [0.5537, 0.6064, 0.9395, 0.6320],
        [0.7568, 0.4140, 0.2257, 0.0815]])

Tensor B:
tensor([[0.8576, 0.6211, 0.5878, 0.2069],
        [0.4574, 0.1433, 0.3266, 0.2308],
        [0.4988, 0.4163, 0.1068, 0.1596]])

A + B:
tensor([[1.0725, 0.6596, 1.3299, 1.0036],
        [1.0112, 0.7497, 1.2661, 0.8627],
        [1.2555, 0.8303, 0.3324, 0.2412]])

A * B:
tensor([[0.1843, 0.0240, 0.4362, 0.1648],
        [0.2533, 0.0869, 0.3068, 0.1458],
        [0.3775, 0.1723, 0.0241, 0.0130]])

A / B:
tensor([[0.2506, 0.0621, 1.2624, 3.8509],
        [1.2106, 4.2326, 2.8768, 2.7388],
        [1.5173, 0.9943, 2.1141, 0.5107]])

Trung bình theo cột của A:
tensor([0.5085, 0.3530, 0.6358, 0.5034])

Độ lệch chuẩn theo cột của A:
tensor([0.2737, 0.2888, 0.3686, 0.3745])


#### Câu 3. Autograd

Cho `y = 2x² + 5x + 3`. Dùng autograd tính `dy/dx` tại `x = 2` và kiểm tra bằng phép tính thủ công.

In [8]:
x = torch.tensor(2.0, requires_grad=True)

y = 2 * x**2 + 5 * x + 3
y.backward()

print("x =", x)
print("y =", y)
print("dy/dx =", x.grad)

x = tensor(2., requires_grad=True)
y = tensor(21., grad_fn=<AddBackward0>)
dy/dx = tensor(13.)


#### Câu 4. Tìm hệ số đa thức

Thay đa thức trong Bài 9 bằng `y = 2x² - 3x + 5`, huấn luyện mô hình và cho biết các hệ số tìm được.

In [9]:
import numpy as np

polynomial = np.poly1d([2, -3, 5])

N = 20
X_data = np.random.randn(N, 1) * 5
Y_data = polynomial(X_data)

XX = np.hstack([
    X_data * X_data,
    X_data,
    np.ones_like(X_data)
])

w = torch.randn(3, 1, requires_grad=True)

x_train = torch.tensor(XX, dtype=torch.float32)
y_train = torch.tensor(Y_data, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(5000):
    optimizer.zero_grad()

    y_pred = x_train @ w
    mse = torch.mean(torch.square(y_train - y_pred))

    mse.backward()
    optimizer.step()

print("\nLearned coefficients:")
print(w)

Initial coefficients:
tensor([[ 0.5770],
        [-2.1052],
        [ 2.1868]], requires_grad=True)

Learned coefficients:
tensor([[ 2.0000],
        [-3.0000],
        [ 4.9999]], requires_grad=True)


#### Nhận xét

- **Câu 1:** Tensor `X` gồm 24 phần tử có thể reshape thành các kích thước `4 × 6`, `2 × 12` và `2 × 3 × 4`. Các phép reshape chỉ thay đổi hình dạng của Tensor mà không làm thay đổi tổng số phần tử.

- **Câu 2:** Hai Tensor `A` và `B` có cùng kích thước `3 × 4` nên có thể thực hiện các phép toán theo từng phần tử như cộng, nhân và chia. Khi sử dụng `dim=0`, trung bình và độ lệch chuẩn được tính theo từng cột.

- **Câu 3:** Với hàm `y = 2x² + 5x + 3`, đạo hàm là `dy/dx = 4x + 5`. Tại `x = 2`, ta có `dy/dx = 13`. Kết quả này trùng với gradient được PyTorch tính bằng autograd.

- **Câu 4:** Các hệ số ban đầu được khởi tạo ngẫu nhiên là `[0.5770, -2.1052, 2.1868]`. Sau 5000 epoch, mô hình học được các hệ số xấp xỉ `[2.0000, -3.0000, 4.9999]`, gần như trùng với các hệ số thực `[2, -3, 5]` của đa thức `y = 2x² - 3x + 5`. Điều này cho thấy autograd kết hợp với optimizer NAdam có thể tối ưu các tham số để tìm lại gần chính xác các hệ số của đa thức.